src/main/java/com/google/sps/servlets/GoServlet.java

```diff
@@ -0,0 +1,316 @@
+   *     the effect of returning multiple types at once.
+   * @param streetAddressWaypointsAsCoordinates A pointer to the result for street address waypoints
+   *     as coordinates to achieve the effect of returning multiple types at once.
+   * @param nonStreetAddressWaypointsAsPlaceTypes A pointer to the result for non street address
+   *     waypoints as place types to achieve the effect of returning multiple types at once.
+   * @throws GeocodingException An exception thrown when an error occurs with the Geocoding API.
+   */
+  public void separateWaypoints(
+      List<String> waypoints,
+      List<String> streetAddressWaypoints,
+      List<Optional<LatLng>> streetAddressWaypointsAsCoordinates,
+      List<Optional<PlaceType>> nonStreetAddressWaypointsAsPlaceTypes)
+      throws GeocodingException {
+    for (String waypoint : waypoints) {
+      GeocodingClient geocodingClient = geocodingClientFactory.getGeocodingClient(apiKey);
+      List<GeocodingResult> geocodingResult = geocodingClient.getGeocodingResult(waypoint);
+      if (GeocodingResultUtility.hasStreetAddress(geocodingResult)) {
+        streetAddressWaypoints.add(waypoint);
+        streetAddressWaypointsAsCoordinates.add(
+            GeocodingResultUtility.getCoordinates(geocodingResult));
+      } else {
+        nonStreetAddressWaypointsAsPlaceTypes.add(
+            GeocodingResultUtility.convertToPlaceType(waypoint));
+      }
+    }
+  }
+
+  /**
+   * Search nearby every street address with known coordinates for a place type match. For example,
+   * if we know exactly where our houses are and we are looking for a restaurant, we search for a
+   * restaurant closest to your house and a restaurant closest to my house. In this case, the method
+   * should return [[restaurantOne, restaurantTwo]]. Scope of method is public for testing purposes.
+   *
+   * @param nonStreetAddressWaypointsAsPlaceTypes A list of place types to call search for. (e.g.
+   *     restaurant, supermarket, police station)
+   * @param streetAddressesAsCoordinates A list of coordinates to look for the place types around.
+   * @return A list of lists of place IDs where each list represents the search nearby result for
+   *     every known coordinate.
+   * @throws PlacesException An exception thrown when an error occurs with the Places API.
+   */
+  public List<List<String>> searchNearbyEveryKnownLocationForClosestPlaceTypeMatch(
+      List<PlaceType> nonStreetAddressWaypointsAsPlaceTypes,
+      List<LatLng> streetAddressesAsCoordinates)
+      throws PlacesException {
+    List<List<String>> allSearchNearbyResults = new ArrayList<>();
+    for (PlaceType nonStreetAddressWaypoint : nonStreetAddressWaypointsAsPlaceTypes) {
+      List<String> searchNearbyResults = new ArrayList<>();
+      for (LatLng coordinate : streetAddressesAsCoordinates) {
+
+        PlacesClient placesClient = placesClientFactory.getPlacesClient(apiKey);
```

Comment:
Sadly, it doesn't work because it needs a new client each time otherwise it throws `java.lang.IllegalStateException: 'await', 'awaitIgnoreError' or 'setCallback' was already called.`. Same with the Geocoding client. :/

Reply:
Gotcha
RESOLVED